### PR TITLE
Type checks response data, stops uncaught type error

### LINKED ISSUE #167: Uncaught TypeError:  Cannot read property '0' of undefined
So we came across this error crashing our app, we cant catch it as its thrown in the lib.

Its caused by a server responding with an unexpected response (I don't actually know if influx responded with it or something else got in the way of the request)

You can repro it from the code here [https://gist.github.com/get-trucked/3336967633e265bfca48954887b8e061](https://gist.github.com/get-trucked/3336967633e265bfca48954887b8e061)

PR to follow soon.

`TypeError: Cannot read property '0' of undefined
    at node_modules/influx/index.js:177:39
    at node_modules/influx/index.js:115:12
    at InfluxRequest._parseCallback (node_modules/influx/lib/InfluxRequest.js:122:10)
    at Request._callback (node_modules/influx/lib/InfluxRequest.js:111:10)
    at Request.self.callback (node_modules/request/request.js:187:22)
    at emitTwo (events.js:87:13)
    at Request.emit (events.js:172:7)
    at Request.<anonymous> (node_modules/request/request.js:1044:10)
    at emitOne (events.js:77:13)
    at Request.emit (events.js:169:7)`

### PR DESCRIPTION
**Pull request by get-trucked**
Fixes #167

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -174,7 +174,16 @@ InfluxDB.prototype.getDatabaseNames = function (callback) {
...
+
+    if (_.isArray(results) && !_.isEmpty(results) && _.isArray(results[0].series) &&
+        !_.isEmpty(results[0].series) && _.isArray(results[0].series[0].values)) {
+
+      return callback(err, _.map(results[0].series[0].values, function (dbarray) {
+        if (_.isArray(dbarray) && !_.isEmpty(dbarray)) {
+          return dbarray[0]
+        }
+      }))
+    } else { return callback(new Error('bad response from server', results))}
```
Comment (connor4312) [inline index.js]: How about something like...

```
var names = _.get(results, '[0].series[0].values')
if (!names) {
  return callback(new Error('bad response from server', results))
}

callback(null, names)
```

Comment (get-trucked) [inline index.js]: I like it, much tidier than mine. 

I still think it needs more type checking as its still easy to crash with a crafted response(I know in normal circumstances this would prob not happen) but you cant catch it outside the lib so I think it should be handled here.

So how about something like this?

```javascript
// These still crash the lib
var results = [{
      series: [{values: 'nope'}]
}]

var results = [{
        series: [{values: [null, undefined]}]
}];
// --------------------------------------------

// More type checking solves above crashes

var names = _.get(results, '[0].series[0].values')
if (!_.isArray(names)) {
  return callback(new Error('bad response from server', results))
}

callback(null, names.map(function (dbarray) {
  if (_.isArray(dbarray)) {
    return dbarray[0]
  }
}))
```

Comment (connor4312) [inline index.js]: Hm, that works, though I'm kind of hesitant about going down the path of guarding against crafted responses. The simplest way to attack the app, which we could not feasibly guard against, would be for an attacker to create an enormous response and DoS the application. If an attacker has the capability to intercept calls you make to the database, you're already pretty screwed.

Comment (get-trucked) [inline index.js]: My intention wasn't to guard from an attacker as if an attacker can intercept a response your well past screwed at that stage. I meant crafted as in my example ones to show the point of should this lib blow up without being able to be caught by the caller.

I haven't read the influx server code but It was to guard from Influx sending bad data (as I'm in not sure if Influx server can send the second response back with an undefined/null as that seemed plausible to me).

Comment (connor4312) [inline index.js]: Seem fair. This lib as it stands right now is a little unstable as far as data handling goes; edge cases sometimes aren't handled, some methods don't work on newer influx versions, etc... If I ever get some free time (maybe possible now that my housing and job situation has settled down!) I'll make a pass through all this code and get things in shape, based on reading of Influx's code and extensive testing of its outputs...

tl;dr sounds good for now! 😄

**Comment by connor4312** [comment]
Thanks!
